Changes to Differential Expression README (#66)
AlexsLemonade/refinebio-examples: differential-expression/README.md

# Refine.bio Example Workflow: Differential expression

[This notebook](https://alexslemonade.github.io/refinebio-examples/differential-expression/gene_DE.nb.html)

> takes data and metadata from refine.bio and identifies differentially expressed genes.

Reviewer: ```suggestion
takes data and metadata from refine.bio and identifies genes that are differentially expressed between two groups.
```
Provided that is accurate

#### RStudio

We have prepared [a quick guide to RStudio](https://github.com/AlexsLemonade/training-modules/blob/master/intro_to_R_tidyverse/00-rstudio_guide.md) as part of our training content that you may find helpful if you're getting started with RStudio for the first time.

Note that the first time you open RStudio, you should select a CRAN mirror.
You can do so by clicking `Tools` > `Global Options` > `Packages` and selecting a CRAN mirror near you with the `Change` button.

#### Interacting with R Notebooks

You can open an R Notebook by opening the `.Rmd` file in RStudio.
Note that working with R Notebooks requires certain R packages, but RStudio should prompt you to download them the first time you open one.
This will allow you to modify and run the R code chunks.
Chunks that have already been included in an example can be run by clicking the green play button in the top right corner of the chunk or by using **Ctrl + Shift + Enter** (**Cmd + Shift + Enter** on a Mac).
See [this guide using to R Notebooks](https://bookdown.org/yihui/rmarkdown/notebook.html#using-notebooks) for more information about inserting and executing code chunks.

> #### Using your own data

Reviewer 2: Could we add this here:
>A statement about downloading in the README (#62 (comment)):
To run the examples yourself, you will need to clone or download this repository < https://github.com/AlexsLemonade/refinebio-examples >

Author: Added that

In order to create a GCT formatted file from a TSV refine.bio data file, download and reference the
[`create_gct_file.R` script](https://github.com/AlexsLemonade/refinebio-examples/blob/master/scripts/create_gct_file.R), followed by `--file` argument with the name of the refine.bio dataset TSV file in your current directory that you would like to convert.
Note: This script requires `optparse` library. If `optparse` library is not installed, this script will install it for you.
Be sure to either have the script and input file in your current working directory, or put type out the full directory path for the script and/or input file. eg. `/users/Bob/Desktop/scripts/create_gct_file.R`
  
If you need more guidance on how to navigate directories, we recommend [this tutorial](https://swcarpentry.github.io/shell-novice/02-filedir/index.html).

Options:     
`--file` :name of the file in your current directory that you would like to convert.    
`--rewrite` :file of the same name as the output will be rewritten (*optional*)       
`--output` :name for the output file, the ".gct" suffix will be added if you do not add it yourself  (*optional*)

> ##### Examples of usage in command line:  

Reviewer: I think this needs more examples that demonstrate what happens if you don't use `--output` and what `--rewrite` is about. The request was for multiple examples.

 ```bash
 $ Rscript scripts/create_gct_file.R \
  --file <PATH TO REFINE.BIO EXPRESSION TSV> \
  --output <PATH TO NEW GCT FILE>
 ```

> Replace the `<PATH TO REFINE.BIO EXPRESSION TSV>` with your file name eg.

Reviewer: ```suggestion
Replace the `<PATH TO REFINE.BIO EXPRESSION TSV>` with your file name, e.g., 
```

`differential-expression/data/HOMO_SAPIENS.tsv`

> Here's an example of the above, where we put the dataset we want to convert in the directory: `differential-expression/data/` and are naming the output file `gct_HOMO_SAPIENS` but saving it to the same directory.

Reviewer: Preference for `HOMO_SAPIENS.gct`.

Also note that for bash commands, a `\` indicates that the command continues on the next line.
Since we put `--rewrite` on the next line, we needed to add a `\` so that it knows that the command continues on the next line.

**Example 4**  
Lastly, if we would like to name the file something besides its original name, we can use the `--output` argument.
Here let's save it directly to the `differential-expression` folder and call it something different.

```bash
Rscript scripts/create_gct_file.R \
 --file differential-expression/data/GSE71270.tsv \
 --output differential-expression/GSE71270_special_name.gct
```

Now if you should see a file called `differential-expression/GSE71270_special_name.gct`.

#### Create a CLS format file

CLS formatted files provide the sample groups or phenotype information and are necessary for performing gene expression differential analysis using GenePattern.
If you've already created a GCT format file from your data, like is described above, you can create a a CLS format using GenePattern's online
[CLSFileCreator](http://software.broadinstitute.org/cancer/software/genepattern/modules/docs/ClsFileCreator/4)

> *Now login into [GenePattern](https://cloud.genepattern.org/gp/pages/login.jsf), select a `Differential Expression` notebook, and follow the instructions to upload and analyze your newly created GCT and CLS files*

Reviewer: This says notebook. It should say module. My comment earlier (https://github.com/AlexsLemonade/refinebio-examples/pull/66#discussion_r280101051) was not about changing this to notebook. In that review, I asked for us to transition to modules.

Author: Oh I see. Got confused.

We suggest saving plots and results to `plots/` and `results/` directories, respectively, as these are automatically created by the notebooks if you move notebooks outside of the GitHub repository structure.
We recommend using the [`limma` user guide](https://www.bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf)
to help you set up your model in a way that takes into account your experimental
set up and hypotheses.

***
## Differential expression analysis with GenePattern

[GenePattern](https://cloud.genepattern.org/gp/pages/login.jsf)
contain ready-made analyses.
For users who are not comfortable with using R Notebooks, the GenePattern modules can be run using a graphics user interface (GUI).
To use GenePattern, you have to create an account.
Here is [their guide](http://software.broadinstitute.org/cancer/software/genepattern/quick-start) we recommend you follow to get started.
For use with GenePattern, data from refine.bio needs to be converted to GenePattern
[formats](http://software.broadinstitute.org/cancer/software/genepattern/file-formats-guide).

> If you would like to do differential expression but would prefer using GenePattern, follow the instructions below.

Reviewer: ```suggestion
If you would like to perform differential expression analysis with refine.bio data but would prefer using GenePattern, follow the instructions below.
```

### Preparing your refine.bio data for GenePattern

In order to complete your differential expression analysis using [GenePattern](https://cloud.genepattern.org/gp/pages/login.jsf), you will need to have:  

>  1) a GCT file    

Reviewer: ```suggestion
 1) a [GCT](https://software.broadinstitute.org/cancer/software/genepattern/file-formats-guide#GCT) file    
```

>  2) a CLS file  

Reviewer: ```suggestion
 2) a [CLS](https://software.broadinstitute.org/cancer/software/genepattern/file-formats-guide#CLS) file  
```

You can follow the steps below to create these files from your refine.bio data.

#### Create a GCT file

Convert a gene expression tab separated values (TSV) file provided into a 'gene cluster text' (GCT) file for use in GenePattern.
In order to create a GCT formatted file from a TSV refine.bio data file, download the
[`create_gct_file.R` script](https://github.com/AlexsLemonade/refinebio-examples/blob/master/scripts/create_gct_file.R).

> To use this script you will need to open [`Terminal`](https://macpaw.com/how-to/use-terminal-on-mac)(for Mac) or [`Command Prompt`](https://www.lifewire.com/command-prompt-2625840)(for Windows).

Reviewer: ```suggestion
To use this script you will need to open [`Terminal`](https://macpaw.com/how-to/use-terminal-on-mac) (for Mac) or [`Command Prompt`](https://www.lifewire.com/command-prompt-2625840) (for Windows).
```

You will need to reference the script like the examples below, followed by `--file` argument with the name of the refine.bio dataset TSV file in your current directory that you would like to convert.
Note: This script requires `optparse` library. If `optparse` library is not installed, this script will install it for you.

**Arguments:**     
`--file`: name of the file in your current directory that you would like to convert.      
`--output`: name for the output file, the ".gct" suffix will be added if you do not add it yourself (*optional*).  
`--rewrite`: file of the same name as the output will be rewritten (*optional*).     

#### Examples of usage in command line:

Below is the basic template for usage of this script.
The following examples will give you an idea of how it works.

```
Rscript scripts/create_gct_file.R \
  --file <PATH TO REFINE.BIO EXPRESSION TSV>
```
To get an idea of how this script and its arguments work, you can run the following examples in order.

> **Go the correct directory**  

Reviewer: ```suggestion
**Navigate to the correct directory**  
```

Depending on where you have put the `refinebio-examples` directory on your computer, you will have to change this path in the code chunk below.

> Be sure to either have the script and input file in your current working directory, or put type out the full directory path for the script and/or input file. eg. `/users/Bob/Desktop/scripts/create_gct_file.R`

Reviewer: ```suggestion
Be sure to either have the script and input file in your current working directory, or type out the full directory path for the script and/or input file. eg. `/users/Bob/Desktop/scripts/create_gct_file.R`
```

For more guidance on how to navigate directories, we recommend [this tutorial](https://swcarpentry.github.io/shell-novice/02-filedir/index.html).

> ```bash

Reviewer: I would either take this chunk or add more context. For example, you'd likely want to point out that this assumes that `refinebio-examples` has been copied to your Desktop.

Author: I will take it out, because the information given in the paragraph *should* help them accomplish this step, but if they blindly put in the code chunk and didn't read the paragraph, they may be confused.

#### Obtaining the R Notebooks

To run the example yourself, you will need to [clone](https://help.github.com/articles/cloning-a-repository/) or [download](https://stackoverflow.com/a/6466993) this repository: https://github.com/AlexsLemonade/refinebio-examples

#### Using your own data

For all the examples in this module, the gene expression data and sample metadata are stored in a `data/` directory.
If you'd like to adapt an example to include data you've obtained from [refine.bio](https://www.refine.bio/), we recommend placing the files in the `data/` directory and changing the filenames and paths in the notebooks to match these files.
We suggest saving plots and results to `plots/` and `results/` directories, respectively, as these are automatically created by the notebooks if you move notebooks outside of the GitHub repository structure.
We recommend using the [`limma` user guide](https://www.bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf)
to help you set up your model in a way that takes into account your experimental
set up and hypotheses.

***

> ## Differential expression analysis with GenePattern

Reviewer: Noted in person: this still says notebooks

```bash
pwd # This will tell you what your current directory is  
cd /Desktop/refinebio-examples  
ls # This will tell you what files exist in the current directory  
```

**Example 1**  
Here we will convert the file `GSE71270.tsv`, which was in our `refinebio-examples/differential-expression/data` directory into a GCT file.
Following the template above, we will replace the `<PATH TO REFINE.BIO EXPRESSION TSV>` with our file name, `differential-expression/data/GSE71270.tsv`

 ```bash
Rscript scripts/create_gct_file.R \
  --file differential-expression/data/GSE71270.tsv
 ```

Note that we have not specified an `--output` name so in this case, the script will use the original name of our file, but replace `.tsv` with `.gct`.
What you should find is that in the same folder `differential-expression/data/`, you now have a file named `GSE71270.gct`

**Example 2**  
After running the code chunk we showed above, let's try running the same thing a second time:

```bash
Rscript scripts/create_gct_file.R \
 --file differential-expression/data/GSE71270.tsv
```

What you should see is an error message that says this:

```

> differential-expression/data/GSE71270.tsv already exists. Use '--rewrite' option if you want this file to be overwritten.

Reviewer: ```suggestion
differential-expression/data/GSE71270.gct already exists. Use '--rewrite' option if you want this file to be overwritten.
```

Correct?

```

This is telling us that `create_gct_file.R` will not write over an already existing file unless we explicitly tell it to.

**Example 3**  
If we want to save over an already existing file, we need to use the `rewrite` option, just like the error message says.  
Let's try that:

```bash
Rscript scripts/create_gct_file.R \
 --file differential-expression/data/GSE71270.tsv \
 --rewrite
```

This will rewrite over the file we made in *Example 1* but should give you a message to tell you it is doing so: `Overwriting file named differential-expression/data/GSE71270.gct`
Also note that for bash commands, a `\` indicates that the command continues on the next line.
Since we put `--rewrite` on the next line, we needed to add a `\` so that it knows that the command continues on the next line.

**Example 4**  
Lastly, if we would like to name the file something besides its original name, we can use the `--output` argument.
Here let's save it directly to the `differential-expression` folder and call it something different.

```bash
Rscript scripts/create_gct_file.R \
 --file differential-expression/data/GSE71270.tsv \
 --output differential-expression/GSE71270_special_name.gct
```

> Now if you should see a file called `differential-expression/GSE71270_special_name.gct`.

Reviewer: ```suggestion
Now you should see a file called `differential-expression/GSE71270_special_name.gct`.
```

See [this guide using to R Notebooks](https://bookdown.org/yihui/rmarkdown/notebook.html#using-notebooks) for more information about inserting and executing code chunks.

#### Using your own data

For all the examples in this module, the gene expression data and sample metadata are stored in a `data/` directory.
If you'd like to adapt an example to include data you've obtained from [refine.bio](https://www.refine.bio/), we recommend placing the files in the `data/` directory and changing the filenames and paths in the notebooks to match these files.
We suggest saving plots and results to `plots/` and `results/` directories, respectively, as these are automatically created by the notebooks if you move notebooks outside of the GitHub repository structure.
We recommend using the [`limma` user guide](https://www.bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf)
to help you set up your model in a way that takes into account your experimental
set up and hypotheses.

***

> ## Alternative differential expression analysis with GenePattern

Author: If I just leave off the word "Alternative" does that idea lessen?

> ## Alternative differential expression analysis with GenePattern

Reviewer: Yes

> [GenePattern notebooks](http://genepattern-notebook.org/example-notebooks/)

Reviewer: Okay - the initial rationale for recommending GenePattern _notebooks_ specifically was a reproducibility angle. If we envision GenePattern as a potential avenue for folks that have less programming experience, then I think we should consider reworking this to talk about and point to the web server modules (the original GenePattern offering). Here's the link to the quick start guide: http://software.broadinstitute.org/cancer/software/genepattern/quick-start

If we put some text around GenePattern versioning (see: http://software.broadinstitute.org/cancer/software/genepattern/concepts#_Version_Numbers), that satisfies the aforementioned reproducibility concern.

##### Example of usage in command line:  
 ```bash
 $ Rscript scripts/create_gct_file.R -f GSE111111.tsv -o outputfilename -r
 ```
Options:     
`-f` :name of the file in your current directory that you would like to convert.    
`-r` :file of the same name as the output will be rewritten (*optional*)       
`-o` :name for the output file (*optional*)       

Optionally you can designate the name of the output file by adding an `-o` argument.
The "GCT" suffix will be added if you do not add it yourself.
Be sure to either have the script and input file in your current working directory, or put type out the full directory path for the script and/or input file. eg. `/users/Bob/Desktop/scripts/create_gct_file.R`

#### Create a CLS format file

CLS formatted files tell information regarding the groups or phenotype of the different samples and are necessary for doing gene expression differential analysis using GenePattern's notebooks.
If you've already created a GCT format file from your data, you can create a a CLS format using GenePattern's online
[CLSFileCreator](http://software.broadinstitute.org/cancer/software/genepattern/modules/docs/ClsFileCreator/4)

> *Now login into [GenePattern](https://cloud.genepattern.org/gp/pages/login.jsf), select a `Differential Expression` module, and follow the instructions to upload and analyze your newly created GCT and CLS files*

Reviewer: This mentions _modules_ rather than _notebooks_.
